### DISCUSSION TITLE
Let "TestClient" support user defined headers

### DISCUSSION THREAD

**Discussion by roy-pstr** [encode/starlette#1961] [Ideas] [answered]
Hi,

I wonder why not let the user pass the "headers" param to TestClient.__init__() function and merge with the default value that is given in super.__init__() ? 

something like:
```
class TestClient(httpx.Client):
...
    def __init__(
        self,
        app: ASGIApp,
        base_url: str = "http://testserver",
        raise_server_exceptions: bool = True,
        root_path: str = "",
        backend: str = "asyncio",
        backend_options: typing.Optional[typing.Dict[str, typing.Any]] = None,
        cookies: httpx._client.CookieTypes = None,
        >> headers: dict = None, <<
    ) -> None:
        ...
        super().__init__(
            app=self.app,
            base_url=base_url,
            >> headers= **headers | **{"user-agent": "testclient"}, <<
            transport=transport,
            follow_redirects=True,
            cookies=cookies,
        )
```
(line changes marked with >> <<)

This feature can be helpful for instance if someone wants to set the authentication headers for a test authenticated client.

Any thoughts here?

**Comment by Kludex** [comment, marked as answer]
Since `cookies` were introduced, maybe we can also include `headers`.

**Reply by roy-pstr** [reply]
Great. 
Send a PR for this?

**Reply by Kludex** [reply]
Hmm... You can create a PR, but I can't promise it's going to be merged. We can use the PR to evaluate better if we want this.

**Reply by danielballan** [reply]
I ran into this same limitation and volunteered to help fix it in https://github.com/encode/starlette/discussions/1914. In case it's useful to you @roy-pstr it is possible to work around this for now by just setting them on the instance:

```py
tc = TestClient(...)
tc.headers.update(headers)
```

but I agree it would be convenient and I think sensible to accept them in `__init__` and pass them up via `super()`.
